duckduckgo/BrowserServicesKit
# Password Hashing

Please review the release process for BrowserServicesKit [here](https://app.asana.com/0/1200194497630846/1200837094583426).
**Required**:
Task/Issue URL: https://app.asana.com/0/1204099484721401/1204378970656376

**Note:** 
This PR is based on and targets the Autofill integration branch.  There are no macOS and iOS PR's required.

What kind of version bump will this require?: Minor

**Optional**:
Tech Design URL: https://app.asana.com/0/481882893211075/1204432765336634
Security Triage: https://app.asana.com/0/1204099484721401/1204432765336635
CC. @bwaresiak 

**Description**:
This updates our WebsiteAccounts model to add a signature hash for comparison purposes in Autofill.   This allows us to remove duplicate credentials from the suggested passwords.   (See associated task and Tech design above)

Specific changes:
- Add a hashing algorithm based on a secure salt stored in the keychain
-  Generate signature on creation and store the has in the account object
- Add a parameter to the account fetching method to filter duplicate credentials
- Add a migration script to update the model and generate hashes for existing user accounts in the vault




**Steps to test this PR**:
Test included.  

Follow these steps to test changes in macOS and iOS

- On macOS, check out the `autofill/subdomain-matching` branch which uses this branch from BSK.
- On iOS point your BSK version to this branch

- Visit [fill.dev/form/login-simple](https://fill.dev/form/login-simple) enter some credentials and save them when offered
- Go to the Autofill UI and add a new entry manually.   Use the same username and password, but `www.fill.dev` as the url
- Go back to the login form and click the Autofill icon in the input text
- **Observe only one account is suggested** 
- Go to the Autofill admin UI and change the username or password for one of the newly created accounts
- Go back to the login form and click the Autofill icon in the input text
- **Observe both accounts are suggested**

![Screen Recording 2023-04-28 at 6 58 12 PM](https://user-images.githubusercontent.com/1156669/235209716-ec80a3dc-60d6-4c93-bcb0-fa41b4142d0e.gif)





---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)
